### PR TITLE
New tokenIds

### PR DESCRIPTION
**Pull request by zajck**
Replaces #582 

Implements necessary changes to support `tokenId = <<uint128(offerId)>>.<<uint128(exchangeId)>>`
Although initially meant only for preminted vouchers, it turned out it's easier to make the change for all new vouchers.

Main reason to do this is to remove binary search in `getPremintStatus`, which has now constant time and space complexity.

Additinal implications:
- max preminted range length was reduced to 2^64 which is still more than anyone realistically needs
- no need to store array of preminted offer ids in vouchers, so ` _rangeOfferIds` was removed. Since binary search was removed, there is no need to require that preminted offer ranges are ordered.
- `transferPremintedFrom` is not needed anymore since regular transfer functions are efficient enough

### CODE REVIEW

**Review by anajuliabit — COMMENTED** [review]
Besides what I put in the code, could you please add unit tests to validate that tokens created before and after EXCHANGE_ID_2_2_0 are burned correctly?

**Review by anajuliabit — CHANGES_REQUESTED** [review]
forgot to choose "request changes"

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 1]
```diff
@@ -816,42 +805,28 @@ contract BosonVoucherBase is
...
+            // it might be a pre-minted token. Preminted tokens have offerId in the upper 128 bits
+            offerId = _tokenId >> 128;
+
+            if (offerId > 0) {
+                // Get the range stored at the midpoint
+                Range storage range = _rangeByOfferId[offerId];
+
+                // Get the beginning of the range once for reference
+                uint256 start = range.start;
+                // _tokenId = _tokenId & type(uint128).max;
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol]: This line should not be here. We need to operate with full tokenIds.
Removed.

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 2]
```diff
@@ -816,42 +805,28 @@ contract BosonVoucherBase is
...
+
+                if (
+                    start > 0 &&
+                    start <= _tokenId &&
+                    start + range.minted - 1 >= _tokenId &&
+                    _tokenId > range.lastBurnedTokenId
+                ) {
+                    // Has it been pre-minted and not burned yet?
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/protocol/clients/voucher/BosonVoucher.sol:828]: Yes `getBosonOffer` is rather costly method. It calls both `getExchange` and `getOffer` on the diamond. And because these are external calls everything needs to be retrieved from storage, copied to memory in send between contracts (around 30 storage slots).

But if you make a dedicated function `getOfferType(uint256 _offerId)` in the protocol that really returns only the type, I think it should be fine.

File: contracts/protocol/facets/ExchangeHandlerFacet.sol [hunk 3]
```diff
@@ -1028,4 +1065,15 @@ contract ExchangeHandlerFacet is IBosonExchangeHandler, BuyerBase, DisputeBase {
             receipt.condition = condition;
         }
     }
+
+    /**
+     * @notice Returns lower 128 bits of the given exchange id.
+     * Needed in case if tokenId is passed in istead of an exchangeId.
```
[the remark this answers is not in the document]
Comment (zajck) [reply contracts/protocol/facets/ExchangeHandlerFacet.sol]: Corrected.

File: test/protocol/OfferHandlerTest.js [hunk 4]
```diff
@@ -319,7 +320,7 @@ describe("IBosonOfferHandler", function () {
       accountId.next(true);
     });
 
-    context("👉 createOffer()", async function () {
+    context.skip("👉 createOffer()", async function () {
```
[the remark this answers is not in the document]
Comment (zajck) [reply test/protocol/OfferHandlerTest.js]: Fixed.

File: test/protocol/OfferHandlerTest.js [hunk 5]
```diff
@@ -1398,7 +1400,7 @@ describe("IBosonOfferHandler", function () {
         );
       });
 
-      context("Owner range is contract", async function () {
+      context.skip("Owner range is contract", async function () {
```
[the remark this answers is not in the document]
Comment (zajck) [reply test/protocol/OfferHandlerTest.js]: Fixed.

**Review by zajck — COMMENTED** [review]
Tests for token burning before/after the upgrade are in #498 .

File: contracts/protocol/clients/voucher/BosonVoucher.sol [hunk 6]
```diff
@@ -199,7 +187,6 @@ contract BosonVoucherBase is
         range.start = _start;
         range.length = _length;
         range.owner = _to;
-        _rangeOfferIds.push(_offerId);
```
Comment (mischat) [inline contracts/protocol/clients/voucher/BosonVoucher.sol:202]: is this related to the above 43 -> 44 change ?

File: scripts/config/protocol-parameters.js [hunk 7]
```diff
@@ -69,4 +69,10 @@ module.exports = {
     mumbai: "0x4102621Ac55e068e148Da09151ce92102c952aab", //dummy
     polygon: "0x17CDD65bebDe68cd8A4045422Fcff825A0740Ef9", //dummy
   },
+
+  EXCHANGE_ID_2_2_0: {
+    hardhat: 1,
+    mumbai: 1, // TODO: adjust for actual deployment
+    polygon: 2, // TODO: adjust for actual deployment
```
Comment (mischat) [inline scripts/config/protocol-parameters.js:76]: do you need to adjust these now?

**Review by mischat — COMMENTED** [review]
I guess, a highlevel comment. The method we added : 'extractExchangeId' states that it is needed when a tokenID is passed in instead of an exchangeID... 

when i look for the usage it only seems to be used in the exchangeHandlderFacet, is this meant to be the case ?

And given that the method, in the comments, states that it takes in a tokenID and outputs an exchangeID, i feel like the naming is misleading : 

```~/Development/bosonprotocol/boson-protocol-contracts/contracts new-tokenId ❯ git grep 'extractExchangeId'                                                                                                                       1m 31s
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:        _exchangeId = extractExchangeId(_exchangeId);
protocol/facets/ExchangeHandlerFacet.sol:    function extractExchangeId(uint256 _exchangeId) internal pure returns (uint256 exchangeId) {```

Looking at the code, we don't seem to use these tokenIDs outside of Voucher and the Exchange handler, every other instance of tokenID seems to either be todo with token gating, token based auth, or twinning, which should be fine ... 

Anyways

**Review by zajck — COMMENTED** [review]
@mischat  Re: extractExchangeId

That was meant more for convenience, so users could pass in either `tokenId` or `exchangeId`.
But yeah, if we really wanted to support that, I would need to do it everywhere (there are methods in dispute and orchestration handler), and also change all natspec to say "exchangeId can either be exchangeId or tokenId`.

So after your comment I changed my mind and removed `extractExchangeId`, since all protocol methods are clear that they expect `exchangeId`. Anyways, whoever is interacting only with the protocol already operates only with the correct exchangeId. 
But even if they have only tokenId, they can always easiliy convert it to exchangeId.

I'm running tests locally now and will push the changes soon.

**Review by anajuliabit — DISMISSED** [review]

**Review by mischat — APPROVED** [review]
Thanks, this looks good to me

**Review by anajuliabit — APPROVED** [review]
LGTM
